autodiff/include/autodiff/TVec.h

```diff
@@ -0,0 +1,287 @@
+template <int DIM>
+TermPtr operator*(const TVec<DIM>& left, const TVec<DIM>& o);
+
+// Helpers:
+template <int DIM>
+inline TVec<DIM> crossProduct(const TVec<DIM>& t1, const TVec<DIM>& t2);
+template <int DIM>
+inline TermPtr distance(const TVec<DIM>& t1, const TVec<DIM>& t2);
+template <int DIM>
+inline TermPtr distanceSqr(const TVec<DIM>& t1, const TVec<DIM>& t2);
+template <int DIM>
+inline TVec<DIM> rotate(const TVec<DIM>& vec, double alpha);
+template <int DIM>
+inline TermPtr projectVectorOntoX(const TVec<DIM>& origin, const TVec<DIM>& dir, TermPtr x);
+template <int DIM>
+inline TVec<DIM> inCoordsOf(const TVec<DIM>& point, const TVec<DIM>& vec);
+template <int DIM>
+inline TermPtr leftOf(const TVec<DIM>& vec, const TVec<DIM>& toCheck);
+template <int DIM>
+inline TermPtr rightOf(const TVec<DIM>& vec, const TVec<DIM>& toCheck);
+template <int DIM>
+inline TermPtr equals(const TVec<DIM>& t1, const TVec<DIM>& t2, double tolerance);
+
+/// Implementation:
+
+template <>
+inline TermPtr TVec<1>::normSquared() const
+{
+    return _terms[0]->getOwner()->constPower(_terms[0], 2);
+}
+
+template <int DIM>
+inline TermPtr TVec<DIM>::normSquared() const
+{
+    TermHolder* h = _terms[0]->getOwner();
+    TermPtr ret = h->constPower(_terms[0], 2.0);
+    for (int i = 1; i < DIM; ++i) {
+        ret = ret + h->constPower(_terms[1], 2.0);
+    }
+    return ret;
+}
+template <int DIM>
+TVec<DIM> TVec<DIM>::normalize() const
+{
+    TermPtr a = normSquared();
+    a = a->getOwner()->constPower(a, 0.5);
+    TVec<DIM> ret;
+    for (int i = 0; i < DIM; ++i) {
+        ret[i] = (_terms[i] / a);
+    }
+    return ret;
+}
+
+template <int DIM>
+TermPtr TVec<DIM>::innerProduct(const TVec<DIM>& o) const
+{
+    TermPtr ret = _terms[0] * o._terms[0];
+    for (int i = 1; i < DIM; ++i) {
+        ret = ret + (_terms[i] * o._terms[i]);
+    }
+    return ret;
+}
+
+template <int DIM>
+TVec<DIM> operator+(const TVec<DIM>& left, const TVec<DIM>& right)
+{
+    TVec<DIM> ret;
+    for (int i = 0; i < DIM; ++i) {
+        ret[i] = left[i] + right[i];
+    }
+    return ret;
+}
+
+template <int DIM>
+TVec<DIM> operator-(const TVec<DIM>& left, const TVec<DIM>& right)
+{
+    TVec<DIM> ret;
+    for (int i = 0; i < DIM; ++i) {
+        ret[i] = left[i] - right[i];
+    }
+    return ret;
+}
+
+template <int DIM>
+TVec<DIM> operator-(const TVec<DIM>& vector)
+{
+    return vector * -1;
+}
+
+template <int DIM>
+TVec<DIM> operator*(const TVec<DIM>& vector, TermPtr scalar)
+{
+    TVec<DIM> ret;
+    for (int i = 0; i < DIM; ++i) {
+        ret[i] = vector[i] * scalar;
+    }
+    return ret;
+}
+
+template <int DIM>
+TVec<DIM> operator*(const TVec<DIM>& vector, const double scalar)
+{
+    return vector * vector[0]->getOwner()->constant(scalar);
+}
+
+template <int DIM>
+TVec<DIM> operator*(TermPtr scalar, const TVec<DIM>& vector)
+{
+    TVec<DIM> ret;
+    for (int i = 0; i < DIM; ++i) {
+        ret[i] = scalar * vector[i];
+    }
+    return ret;
+}
```

Comment:
I believe that since this is the * operator, you want to give the user of this library as much flexibility as possible to express their problems how they want. So, this allows the user to specify either a vector or a scalar on the left or right side of the * operator.

Reply:
Yes, basically allow 3*vec and vec*3